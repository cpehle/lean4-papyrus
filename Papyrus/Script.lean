import Papyrus.Script.Module
import Papyrus.Script.Function
import Papyrus.Script.Instructions
import Papyrus.Script.Types
