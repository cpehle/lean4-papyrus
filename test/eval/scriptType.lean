import Papyrus.Script.Type

open Papyrus Script

-- # Floating Point Types

#check llvm type half
#check llvm type bfloat
#check llvm type float
#check llvm type double
#check llvm type x86_fp80
#check llvm type fp128
#check llvm type ppc_fp128

-- # Special Types

#check llvm type void
#check llvm type label
#check llvm type metadata
#check llvm type x86_mmx
#check llvm type x86_amx
#check llvm type token

-- # Integer Types

#check i1
#check i32
#check i1942652

#check llvm type i1
#check llvm type i32
#check llvm type i1942652

-- # Function Types

#check llvm type i32 (i32)
#check llvm type float (i16, i32 *) *
#check llvm type i32 (i8*, ...)
#check llvm type {i32, i32} (i32)

-- # Pointer Types

#check i8*

#check llvm type i8*
#check llvm type [4 x i32]*
#check llvm type i32 (i32*) *
#check llvm type i32 addrspace(5) *

-- # (Literal) Struct Types

#check llvm type { i32, i32, i32 }
#check llvm type { float, i32 (i32) * }
#check llvm type <{ i8, i32 }>

-- # Array Types

#check llvm type [40 x i32]
#check llvm type [41 x i32]
#check llvm type [4 x i8]

#check llvm type [3 x [4 x i32]]
#check llvm type [12 x [10 x float]]
#check llvm type [2 x [3 x [4 x i16]]]

-- # Vector Types

#check llvm type <4 x i32>
#check llvm type <8 x float>
#check llvm type <2 x i64>
#check llvm type < 4 x i64* >
#check llvm type <vscale x 4 x i32>

-- # Nested Terms

#check llvm type [4 Ã— type(int8Type)]
#check llvm type { type(int8Type), type(floatType) }

-- #check llvm type <4 x i64*> -- fails: `*>` is a separate token

-- #check llvm type %X
-- #check llvm type %T1 { i32, i32, i32 }
-- #check llvm type %T2 <{ i8, i32 }>
