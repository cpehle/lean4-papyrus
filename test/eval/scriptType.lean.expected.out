halfType : Type
bfloatType : Type
floatType : Type
doubleType : Type
x86FP80Type : Type
fp128Type : Type
ppcFP128Type : Type
voidType : Type
labelType : Type
metadataType : Type
x86MMXType : Type
x86AMXType : Type
tokenType : Type
integerType 1 : IntegerType
integerType 32 : IntegerType
integerType 1942652 : IntegerType
integerType 1 : IntegerType
integerType 32 : IntegerType
integerType 1942652 : IntegerType
functionType (Type.integer (integerType 32)) #[Type.integer (integerType 32)] : FunctionType
(Type.function
    (functionType floatType
      #[Type.integer (integerType 16), Type.pointer (Type.integer (integerType 32))*]))* : PointerType
functionType (Type.integer (integerType 32)) #[Type.pointer (Type.integer (integerType 8))*] true : FunctionType
functionType (Type.struct (literalStructType #[Type.integer (integerType 32), Type.integer (integerType 32)]))
  #[Type.integer (integerType 32)] : FunctionType
(Type.integer (integerType 8))* : PointerType
(Type.integer (integerType 8))* : PointerType
(Type.array (arrayType (Type.integer (integerType 32)) 4))* : PointerType
(Type.function
    (functionType (Type.integer (integerType 32)) #[Type.pointer (Type.integer (integerType 32))*]))* : PointerType
pointerType (Type.integer (integerType 32)) 5 : PointerType
literalStructType
  #[Type.integer (integerType 32), Type.integer (integerType 32), Type.integer (integerType 32)] : StructType
literalStructType
  #[floatType,
    Type.pointer
      (Type.function (functionType (Type.integer (integerType 32)) #[Type.integer (integerType 32)]))*] : StructType
literalStructType #[Type.integer (integerType 8), Type.integer (integerType 32)] true : StructType
arrayType (Type.integer (integerType 32)) 40 : ArrayType
arrayType (Type.integer (integerType 32)) 41 : ArrayType
arrayType (Type.integer (integerType 8)) 4 : ArrayType
arrayType (Type.array (arrayType (Type.integer (integerType 32)) 4)) 3 : ArrayType
arrayType (Type.array (arrayType floatType 10)) 12 : ArrayType
arrayType (Type.array (arrayType (Type.array (arrayType (Type.integer (integerType 16)) 4)) 3)) 2 : ArrayType
fixedVectorType (Type.integer (integerType 32)) 4 : FixedVectorType
fixedVectorType floatType 8 : FixedVectorType
fixedVectorType (Type.integer (integerType 64)) 2 : FixedVectorType
fixedVectorType (Type.pointer (Type.integer (integerType 64))*) 4 : FixedVectorType
scalableVectorType (Type.integer (integerType 32)) 4 : ScalableVectorType
arrayType (Type.integer int8Type) 4 : ArrayType
literalStructType #[Type.integer int8Type, floatType] : StructType
